File: internal/agent/binary_identity.go
package agent

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
)

const binaryIdentityFileName = "binary-id"

func binaryIdentityPath(paths Paths) string {
	if strings.TrimSpace(paths.SupportDir) == "" {
		return ""
	}
	return filepath.Join(paths.SupportDir, binaryIdentityFileName)
}

func binaryIdentityForExecutable(path string) (string, error) {
	cleaned := filepath.Clean(strings.TrimSpace(path))
	if cleaned == "" {
		return "", fmt.Errorf("missing executable path")
	}

	file, err := os.Open(cleaned)
	if err != nil {
		if os.IsNotExist(err) {
			return "path:" + cleaned, nil
		}
		return "", fmt.Errorf("open executable %s: %w", cleaned, err)
	}
	defer file.Close()

	hash := sha256.New()
	if _, err := io.Copy(hash, file); err != nil {
		return "", fmt.Errorf("hash executable %s: %w", cleaned, err)
	}
	return "sha256:" + hex.EncodeToString(hash.Sum(nil)), nil
}

func readBinaryIdentity(path string) (string, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(data)), nil
}

func writeBinaryIdentity(path, identity string) error {
	if strings.TrimSpace(path) == "" {
		return fmt.Errorf("missing binary identity path")
	}
	if err := os.WriteFile(path, []byte(strings.TrimSpace(identity)+"\n"), 0o600); err != nil {
		return fmt.Errorf("write binary identity %s: %w", path, err)
	}
	return nil
}

File: Sources/XcodeCLICore/Agent/BinaryIdentity.swift
import Foundation
import CryptoKit

private let binaryIdentityFileName = "binary-id"

/// Compute the path to the binary identity file.
public func binaryIdentityPath(_ paths: AgentPaths.Paths) -> String {
    guard !paths.supportDir.trimmingCharacters(in: .whitespaces).isEmpty else { return "" }
    return (paths.supportDir as NSString).appendingPathComponent(binaryIdentityFileName)
}

/// Compute the identity string for an executable.
/// Returns "sha256:<hex>" if the file exists, "path:<absolute_path>" otherwise.
public func binaryIdentityForExecutable(_ path: String) throws -> String {
    let cleaned = (path.trimmingCharacters(in: .whitespaces) as NSString).standardizingPath
    guard !cleaned.isEmpty else {
        throw XcodeCLIError.bridgeSpawnFailed(underlying: "missing executable path")
    }

    guard let data = FileManager.default.contents(atPath: cleaned) else {
        if !FileManager.default.fileExists(atPath: cleaned) {
            return "path:\(cleaned)"
        }
        throw XcodeCLIError.bridgeSpawnFailed(underlying: "cannot read executable \(cleaned)")
    }

    let hash = SHA256.hash(data: data)
    let hex = hash.map { String(format: "%02x", $0) }.joined()
    return "sha256:\(hex)"
}

/// Read the persisted binary identity.
public func readBinaryIdentity(_ path: String) throws -> String {
    try String(contentsOfFile: path, encoding: .utf8).trimmingCharacters(in: .whitespacesAndNewlines)
}

/// Write a binary identity string to file.
public func writeBinaryIdentity(_ path: String, identity: String) throws {
    guard !path.trimmingCharacters(in: .whitespaces).isEmpty else {
        throw XcodeCLIError.bridgeSpawnFailed(underlying: "missing binary identity path")
    }
    try (identity.trimmingCharacters(in: .whitespaces) + "\n")
        .write(toFile: path, atomically: true, encoding: .utf8)
}
